### ISSUE TITLE
[Bug]: Unexpected Dependency false when importing component inside MDX file

### ISSUE THREAD

**Issue by Yonben** [storybookjs/storybook#20958]
### Describe the bug

When writing my story as MDX, I `import * as ButtonStories` but also import the `Button` component itself to use it in some Doc components I build. It leads to `Unexpected Dependency: false` error.

Is there a way I can import my component to the MDX without breaking that dep? Is the error somewhere else?

### To Reproduce

```shell
https://codesandbox.io/p/sandbox/romantic-shockley-ghnrvc
```

### System

```shell
Environment Info:

  System:
    OS: Linux 5.15 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
    CPU: (2) x64 AMD EPYC
  Binaries:
    Node: 16.17.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.15.0 - /usr/local/bin/npm
  npmPackages:
    @storybook/addon-essentials: 7.0.0-beta.43 => 7.0.0-beta.43 
    @storybook/addon-interactions: 7.0.0-beta.43 => 7.0.0-beta.43 
    @storybook/addon-links: 7.0.0-beta.43 => 7.0.0-beta.43 
    @storybook/addons: 7.0.0-beta.43 => 7.0.0-beta.43 
    @storybook/react: 7.0.0-beta.43 => 7.0.0-beta.43 
    @storybook/react-vite: 7.0.0-beta.43 => 7.0.0-beta.43
```

**Comment by valentinpalkovic** [comment]
@tmeasday, @JReinhold

**Comment by Yonben** [comment]
@JReinhold and me took a look. He thought maybe the `import { Button } from './Button';` was causing an issue since it could refer to `Button.tsx` and `Button.mdx`. So I updated the repro sandbox by changing the MDX file to `ButtonDocs.mdx` but the issue still shows:
```
@storybook/cli v7.0.0-beta.43

info => Starting manager..
info => Serving static files from ././node_modules/@storybook/manager/static at /sb-manager-assets
WARN 🚨 Extraction error on src/components/Button/ButtonDocs.mdx: Error: Unexpected dependency: false
WARN Force closed manager build
ERR! Error: Unexpected dependency: false
```

Thanks!

**Comment by Yonben** [comment]
More progress. Adding the extension explicitly `import { Button } from './Button.tsx';` works. Doesn't make sense since there are no other `Button` file to resolve, and imports of other components work without extension.

TLDR of our talk with @JReinhold:
Considering the rest is documented as "regular without extension" imports and suddenly it has to be different there, it feels uncomfortable if it makes sense.
We should at least understand clearly why this behavior happens.
And then if the extension is necessary, document that, but first understand it.

Thanks again ✨

**Comment by tmeasday** [comment]
@shilman that's how I read it too, but I think it's the other way round:

The keys of the `cache`s get set here (based on globs from `main.js`):

https://github.com/storybookjs/storybook/blob/e5cded85e71953d0cb5fcdfec2997ebbdc768674/code/lib/core-server/src/utils/StoryIndexGenerator.ts#L133

So if we are filtering the keys to find the referenced file names, we would simply find keys that match stories globs.

The only reason a cache entry would exist and be `false` would be if it was not processed yet, i.e another docs file. It's all very confusing.

**Comment by tmeasday** [comment]
Ok, thanks to some screensharing we worked out that the file that it is matching (incorrectly) is itself (`./Button` being a prefix of `./ButtonDocs.mdx`). 

The prefix part is a bit of a bug but ultimately it could have been importing from `./ButtonDocs` (`.tsx`) and had the same issue. There's no easy way for us to tell what it's matching; but if we search through the index and find a docs or unprocessed file (which by definition is docs) we should just skip it.

**Comment by shilman** [comment]
Huzzah!! I just released https://github.com/storybookjs/storybook/releases/tag/v7.0.0-beta.61 containing PR #21348 that references this issue. Upgrade today to the `@next` NPM tag to try it out!

```
npx sb@next upgrade --prerelease
```
